### PR TITLE
Update WT's icon at runtime to match high-contrast as applicable

### LINKED ISSUE #6777: Terminal Preview and Terminal have the same icon in Alt-Tab and Win-Tab window switchers
### Environment

**Windows build number:** Windows 10 ver. 2004 build 19041.329
**Windows Terminal version (if applicable):** 1.0.1811.0, Preview 1.1.1812.0

### Steps to reproduce

1. Open Windows Terminal.
1. Open Windows Terminal Preview.
1. Press Alternate-Tab or Windows-Tab.

### Expected behavior

The Terminal and Terminal Preview windows thumbnails have different icons displayed next to their titles.

### Actual behavior

The Terminal and Terminal Preview windows thumbnails have the same displayed next to their titles.
(Taskbar shows different icons as expected.)
Windows-Tab screenshot showing Windows Terminal and Windows Terminal Preview with the same icons

### PR DESCRIPTION
**Pull request by DHowett**
This commit introduces 8 more variants of the .ICO file, embeds the
right ones into WindowsTerminal.exe, and adds code that will select the
most appropriate icon at runtime.

Since we're a Centennial application, the "application" icon inside our
package isn't used by the shell for the taskbar thumbnails or the
Alt-Tab window.

To quote J. Tippet,
> I believe there are two possible fixes:
>
> 1. Fix the OS shell to prefer the MRT icon instead of preferring the
>    win32 icon
> 2. Add alternate versions of /res/terminal.ico
> The 1st fix is clearly better, since it benefits any hybrid app. But
> the 2nd fix is much easier, since it'd just take about an hour to gin up
> a new .ico file and hack the .RC file to refer to it when building the
> preview flavor.

... and to quote Michael Ratanapintha,

> Basically, if your MSIX-packaged desktop app's image resources are
> separate files or even separate MSIX packages, they may be loaded by
> MRT. If they're embedded in the .exe, they're the old-fashioned Win32
> resources Mr. Tippet is referring to.

This is the "2nd fix."

Fixes #6777

Co-authored-by: Jeffrey Tippet <jtippet@ntdev.microsoft.com>

### CODE REVIEW

File: src/cascadia/WindowsTerminal/icon.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    const auto found{ std::find_if(s_iconClassMapping.cbegin(), s_iconClassMapping.cend(), [&](auto&& entry) { return entry.first == iconClass; }) };
+    const auto resource{ found != s_iconClassMapping.cend() ? found->second : IDI_APPICON };
+    return MAKEINTRESOURCEW(resource);
+}
+
+void UpdateWindowIconForActiveMetrics(HWND window)
+{
+    auto iconResource{ _GetActiveAppIcon() };
+    HANDLE smallIcon{ LoadImageW(wil::GetModuleInstanceHandle(), iconResource, IMAGE_ICON, GetSystemMetrics(SM_CXSMICON), GetSystemMetrics(SM_CYSMICON), LR_DEFAULTCOLOR) };
+    HANDLE largeIcon{ LoadImageW(wil::GetModuleInstanceHandle(), iconResource, IMAGE_ICON, GetSystemMetrics(SM_CXICON), GetSystemMetrics(SM_CYICON), LR_DEFAULTCOLOR) };
```
Comment (DHowett) [inline src/cascadia/WindowsTerminal/icon.cpp:73]: leaking handles, but rarely

**Comment by metathinker** [comment]
>To quote J. Tippet (@jtippet),
>... and to quote M. Ratanapintha (@metathinker),

I don't know how Jeff likes to be called, but I personally prefer to be called just Michael, thank you very much 🙂

**Comment by DHowett** [comment]
@metathinker uniformity defeated common sense. Thanks!

**Comment by metathinker** [comment]
I like both words of my name. I find it easier to teach people how to spell and pronounce my first name, though :)

File: src/cascadia/WindowsTerminal/icon.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    HIGHCONTRASTW hcInfo{};
+    hcInfo.cbSize = sizeof(hcInfo);
+
+    if (SystemParametersInfoW(SPI_GETHIGHCONTRAST, sizeof(hcInfo), &hcInfo, 0))
+    {
+        if (WI_IsFlagSet(hcInfo.dwFlags, HCF_HIGHCONTRASTON))
+        {
+            WI_SetFlag(iconClass, IconClass::Variant_HC);
+
+            if (hcInfo.lpszDefaultScheme)
```
Comment (jtippet) [inline src/cascadia/WindowsTerminal/icon.cpp:61]: i'm sure you know this doesn't localize well.  instead, you can use this approach:

```
if (0x00FFFFFF == GetSysColor(COLOR_WINDOW)) {
    WI_SetFlag(HC_White);
}
```
(internally, grep for SystemThemingInterop::GetSystemHighContrastTheme)

I'm not sure if XAML theme resources are available in this part of the application, but if they are, you could alternatively let the XAML theming decide the right theme for you:
```
            <ResourceDictionary.ThemeDictionaries>
                <ResourceDictionary x:Key="Default">
                    <x:Int32 x:Key="Win32IconResourceId">101</x:Int32>
                </ResourceDictionary>
                <ResourceDictionary x:Key="HighContrastBlack">
                    <x:Int32 x:Key="Win32IconResourceId">102</x:Int32>
                </ResourceDictionary>
                <ResourceDictionary x:Key="HighContrastWhite">
                    <x:Int32 x:Key="Win32IconResourceId">103</x:Int32>
                </ResourceDictionary>
            </ResourceDictionary.ThemeDictionaries>
        </ResourceDictionary>
```
😹

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 3]
```diff
@@ -147,6 +148,8 @@ void IslandWindow::_HandleCreateWindow(const WPARAM, const LPARAM lParam) noexce
     ShowWindow(_window.get(), nCmdShow);
 
     UpdateWindow(_window.get());
+
+    UpdateWindowIconForActiveMetrics(_window.get());
```
Comment (jtippet) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:152]: did you also mean to call this from WM_THEMECHANGED ?

File: src/cascadia/WindowsTerminal/resource.h [hunk 4]
```diff
@@ -3,20 +3,31 @@
...
-#define IDS_X86_ARCHITECTURE 111
-#define IDS_AMD64_ARCHITECTURE 112
-#define IDS_ARM64_ARCHITECTURE 113
-#define IDS_ARM_ARCHITECTURE 114
-#define IDS_UNKNOWN_ARCHITECTURE 115
+#define IDI_APPICON_HC_B 102
+#define IDI_APPICON_HC_W 103
+
+#define IDI_APPICON_DEV 104
+#define IDI_APPICON_DEV_HC_B 105
```
Comment (jtippet) [inline src/cascadia/WindowsTerminal/resource.h:10]: so i am just now noticing that the DEV_HC icons do not actually have the "DEV" badge, even though the SVG has it.  that appears to be a bug left over from when i monkeyed with the icons.  i just tried running the script, and the icons got better... i wonder how i overlooked doing this?  but can i request that you slip this in, while you're at it?

```
.\Generate-TerminalAssets.ps1 -path .\Terminal_Dev.svg -HighContrastPath .\Terminal_Dev_HC.svg -Destination .\images-dev
```

Comment (jtippet) [inline src/cascadia/WindowsTerminal/resource.h:10]: btw, maybe also update .\Generate-TerminalAssets.ps1 to automate the creation of the win32 assets (terminal-Pre_contrast-black.ico et al).

File: src/cascadia/WindowsTerminal/WindowsTerminal.rc [hunk 5]
```diff
@@ -53,6 +53,16 @@ END
 // Icon with lowest ID value placed first to ensure application icon
 // remains consistent on all systems.
 IDI_APPICON             ICON                    "..\\..\\..\\res\\terminal.ico"
```
Comment (jtippet) [inline src/cascadia/WindowsTerminal/WindowsTerminal.rc:55]: this feels like the sort of thing that would be appropriate to use #ifdefs on -- there's no need to ship all icons in all variants, and no need to spend CPU at runtime to compute the build variant (imho).  does your project system expose any preprocessor macros for this?

i'm suggesting something like

```
#if OFFICIAL_BUILD
IDI_APPICON ICON "res\\terminal.ico"
IDI_APPICON_HC ICON "res\\terminal_contrast-black.ico"
#elif PREVIEW_BUILD
IDI_APPICON ICON "res\\terminal-Pre.ico"
IDI_APPICON_HC ICON "res\\terminal-Pre_contrast-black.ico"
#else
IDI_APPICON ICON "res\\terminal-Dev.ico"
IDI_APPICON_HC ICON "res\\terminal-Dev_contrast-black.ico"
#endif
```

**Comment by jtippet** [comment]
> I don't know how Jeff likes to be called, but I personally prefer to be called just Michael, thank you very much 🙂

To avoid ambiguity with other Jeffs, I prefer to be called 272fc969-1ffe-4e01-9589-5914c08d1e9a.

**Comment by DHowett** [comment]
This explains a lot.

image

File: src/cascadia/WindowsTerminal/IslandWindow.cpp (hunk 3 above)
Comment (DHowett) [reply src/cascadia/WindowsTerminal/IslandWindow.cpp:152]: I did, but I chose to defer that until we also get WM_SETTINGSCHANGED.

File: src/cascadia/WindowsTerminal/icon.cpp (hunk 2 above)
Comment (DHowett) [reply src/cascadia/WindowsTerminal/icon.cpp:61]: I wish we had access to theme resources here, but alas.. we don't. ☹️

Good call with the window color. I prefer that over the theme name hack.

File: src/cascadia/WindowsTerminal/resource.h (hunk 4 above)
Comment (DHowett) [reply src/cascadia/WindowsTerminal/resource.h:10]: Sure! Good catch.

File: src/cascadia/WindowsTerminal/WindowsTerminal.rc (hunk 5 above)
Comment (DHowett) [reply src/cascadia/WindowsTerminal/WindowsTerminal.rc:55]: So, right now Terminal _doesn't build different variants_ for its different packages.

Our original intent with this build tooling was that a single build would produce both packages from the same assets such that our ring promotion engineer (so, me) could take the same package and stage it for both Preview and Stable.

Given that we re-spin every stable build anew off backported changes from the last preview, I'm not sure we're benefitting from that, but I'm not sure this is the place to change that. I'll happily hold this change back until I do that other thing.

File: src/cascadia/WindowsTerminal/resource.h [hunk 6]
```diff
@@ -3,20 +3,31 @@
...
+#define IDI_APPICON_PRE_HC_W 109
+
+#define IDS_ERROR_DIALOG_TITLE 110
+#define IDS_HELP_DIALOG_TITLE 111
+#define IDS_ERROR_ARCHITECTURE_FORMAT 112
+#define IDS_X86_ARCHITECTURE 113
+#define IDS_AMD64_ARCHITECTURE 114
+#define IDS_ARM64_ARCHITECTURE 115
+#define IDS_ARM_ARCHITECTURE 116
+#define IDS_UNKNOWN_ARCHITECTURE 117
```
[the remark this answers is not in the document]
Comment (DHowett) [reply src/cascadia/WindowsTerminal/resource.h:24]: I was wondering about that! Thanks!

**Comment by jtippet** [comment]
So, uh, someone wrote [a linting tool](https://github.com/jtippet/icotools#iconag) for .ICO files, and it really doesn't like your .ICOs.  I guess ImageMagick isn't magic at making .ICOs?

```
PS C:\> dotnet IcoInfo.dll -i images\terminal_contrast-white.ico
File: images\terminal_contrast-white.ico
terminal_contrast-white.ico(1): Warning ICO241: ICO files require the embedded PNG image to be encoded in RGBA32 format; this is GrayscaleAlpha
  Frame #1
    Encoding:      PNG
    Bytes on disk: 1886
                   Actual     Claimed in ICO header
    Width:            256                       256
    Height:           256                       256
    Bit depth:         16                        32

terminal_contrast-white.ico(2): Error ICO131: BITMAPINFOHEADER.biXPelsPerMeter should be 0, was 5038.
More information online:
    ICO131: https://github.com/jtippet/IcoTools/wiki/ICO131
    ICO241: https://github.com/jtippet/IcoTools/wiki/ICO241
```

.ICO files can contain multiple images, which are differentiated by <Height, Width, ColorDepth>.  (These days, color depth is obsolete and you can just use 32bpp everywhere.)  When you call LoadImageW on an icon resource, the loader tries to find a perfect match for the size & depth.  If it can't, it finds the image that's heuristically "closest", then uses a crappy scaling algorithm to resize it.

Rather than (a) paying the CPU cost of resizing at runtime, (b) paying the memory cost of having a bitmap in memory versus one that's just on disk, and (c) letting win32k's crappy scaler handle your bitmap, it's really better to ship an ICO bundled with prepared image sizes that you expect to use.  (In this way, a .ICO is a lot like a MRT bundle, except worse in pretty much every way).

I think you're *trying* to do this right, since the .PS1 script mentions a variety of win32 asset sizes.  But whatever ImageMagick is doing... you wound up with a .ICO that has a single 256x256 image in it.  That's pretty much guaranteed to need to be scaled down whenever it's used.

As an additional wrinkle: any individual image in a .ICO can either be stored as a .BMP file or as a .PNG file.  But in both cases, .ICO only tolerates a kooky subset of the respective image formats.  For PNG, it only supports RGBA32 encoding; you can't just use any old PNG encoding.  This gotcha will getcha bad, because the OS has at least two separate .ICO decoders: one in user32, and another in WIC.  The user32 decoder tolerates any valid PNG file, but WIC will reject your image if it's not exactly RGBA32.  So your icon will be broken in some GUI surfaces, depending on whether that GUI uses user32 or WIC to draw icons.

**Comment by DHowett** [comment]
This is strange. VS and other tools are showing all 6 icon variants in the .ico file. Shell handles it well, too. That's quite curious.

**Comment by DHowett** [comment]
Sorry-- by "well" I mean that shell uses the 16x16 variant in the 16x16 surfaces.

**Comment by jtippet** [comment]
Oh so it is... that's because the tool is bailing out when it sees the next error
```
terminal_contrast-white.ico(2): Error ICO131: BITMAPINFOHEADER.biXPelsPerMeter should be 0, was 5038. 
```

**Comment by jtippet** [comment]
Ok, let me retract most of the commentary about putting multiple images in there.  I'll stick with the comment about the 256x256 tile though: that needs to be converted to RGBA32, or else the 256px icon won't show up in any WIC-rasterized surface.

**Comment by DHowett** [comment]
(your icon toolset is _very cool_. Do you want me to snag the broken files from this commit and file an issue titled, generally, "this bad icon file makes IcoInfo upset"?)

**Comment by jtippet** [comment]
I already reported the bug to ImageMagick, who were amazingly quick to respond: https://github.com/ImageMagick/ImageMagick/issues/2756

Meanwhile, I know you didn't expect to spend multiple days on this PR, and I acknowledge that much of that is noise from me.  So I've generated compliant icons here, if you want to just cherry-pick them: https://github.com/jtippet/terminal/commit/5c2e61100ea21f4bfb569d4451e3b567a37be6e6

They were created as the output of:
```
del terminal.ico
icocat -i terminal.ico -s _intermediate.standard.256.png
icocat -i terminal.ico -s _intermediate.standard.64.png
. . . all other sizes . . .
icocrush -i terminal.ico --write-in-place --preferred-format PngLargeImages --emit-masked-image-pixel Compliant
```

They contain a structure like this, which resembles the original terminal.ico:
```
File: images-Pre\terminal_contrast-white.ico
  Frame #1
    Encoding:      PNG
    Bytes on disk: 9814
                   Actual     Claimed in ICO header
    Width:            256                       256
    Height:           256                       256
    Bit depth:         32                        32

  Frame #2
    Encoding:      Bitmap
    Bitmap type:   32-bit ARGB
    Bytes on disk: 16952
                   Actual     Claimed in ICO header
    Width:             64                        64
    Height:            64                        64
    Bit depth:         32                        32

  Frame #3
    Encoding:      Bitmap
    Bitmap type:   32-bit ARGB
    Bytes on disk: 9656
                   Actual     Claimed in ICO header
    Width:             48                        48
    Height:            48                        48
    Bit depth:         32                        32

  Frame #4
    Encoding:      Bitmap
    Bitmap type:   32-bit ARGB
    Bytes on disk: 4280
                   Actual     Claimed in ICO header
    Width:             32                        32
    Height:            32                        32
    Bit depth:         32                        32

  Frame #5
    Encoding:      Bitmap
    Bitmap type:   32-bit ARGB
    Bytes on disk: 1736
                   Actual     Claimed in ICO header
    Width:             20                        20
    Height:            20                        20
    Bit depth:         32                        32

  Frame #6
    Encoding:      Bitmap
    Bitmap type:   32-bit ARGB
    Bytes on disk: 2456
                   Actual     Claimed in ICO header
    Width:             24                        24
    Height:            24                        24
    Bit depth:         32                        32

  Frame #7
    Encoding:      Bitmap
    Bitmap type:   32-bit ARGB
    Bytes on disk: 1144
                   Actual     Claimed in ICO header
    Width:             16                        16
    Height:            16                        16
    Bit depth:         32                        32
```

## Needless tangent
There's kind of an interesting tradeoff on whether to use BMP or PNG for any particular image.  In this case, the PNG is _always_ smaller than the BMP, even for the 16x16 image.  So if you say "shrink the disk, at all costs", you'd encode everything as PNG.  But there's a couple downsides to that:
* Windows XP doesn't support PNG in ICOs, so you'll confuse anyone running ... that.
* PNGs take more CPU to decode and display.

So I decided to encode the 256x256 as PNG, and everything smaller as BMP.  I chose this heuristic because that's what your original terminal.ico does ;).  If you were to do PNGs everywhere, it'd bring the filesize from 43kb to 12kb, which is kind of surprising, but not worth forcing someone to load up a PNG decoder just to look at your app's icon.

**Comment by DHowett** [comment]
I seriously appreciate that. Thank you.

PNGs for 256x256 and BMPs for everything else is totally fine with me.

**Review by zadjii-msft — APPROVED** [review]
This is pretty neat, not gonna lie

**Review by miniksa — APPROVED** [review]

**Pull request merged**